### DISCUSSION TITLE
Number of Legal Moves to empty squares only

### DISCUSSION THREAD

**Discussion by infinitless** [niklasf/python-chess#846] [Q&A] [answered]
Is there a way to separate legal moves to empty squares from legal moves that are captures? And count both?

**Comment by niklasf** [comment, marked as answer]
You can use bitboard masks to achieve this.

```python
gen = board.generate_legal_moves(to_mask=~board.occupied)  # To empty squares only
gen = board.generate_legal_moves(to_mask=board.occupied)   # To occupied squares only
```

Note that en passant captures are moves to empty squares. Use `board.generate_legal_captures()` to include these.

These functions return generators. Moves can be counted using `len(list(gen))` or `sum(1 for _ in gen)`.

**Reply by infinitless** [reply]
Great. Thanks!

**Reply by cherepanovic** [reply]
what is the way to get from the gen (from above) to SquareSet?
